### PR TITLE
send docfx config event

### PR DESCRIPTION
**Pull request by 928PJY**
[AB#264062](https://dev.azure.com/ceapex/d3d54af3-265a-4f18-95f6-9a46397ca583/_workitems/edit/264062)

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/docfx/pull/6216)

### CODE REVIEW

File: src/docfx/lib/log/DocfxBuildTelemetryModel.cs [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+{
+    [JsonObject(NamingStrategyType = typeof(SnakeCaseNamingStrategy))]
+    internal class DocfxBuildTelemetryModel
+    {
+        public Dictionary<string, Dictionary<string, object>> FileMetadata { get; set; } = new Dictionary<string, Dictionary<string, object>>();
+
+        [JsonExtensionData]
+        public Dictionary<string, object> ExtensionData { get; set; } = new Dictionary<string, object>();
+
+        public bool ShouldSerializeFileMetadata() => false;
```
Comment (yufeih) [inline src/docfx/lib/log/DocfxBuildTelemetryModel.cs:18]: trim complex leaf nodes when size exceed?

File: src/docfx/lib/log/Telemetry.cs [hunk 2]
```diff
@@ -73,6 +74,28 @@ public static void SetRepository(string? repo, string? branch)
...
+            }
+            catch (Exception ex)
+            {
+                Console.WriteLine($"Error happens when deserializing the config file {fileName}: {ex.Message}");
+            }
+
+            var dimensions = new Dictionary<string, string>();
+            dimensions["DocsetName"] = docsetName;
+            dimensions["Config"] = docfxConfigTelemetryValue;
+            TrackEvent(fileName, dimensions);
```
Comment (yufeih) [inline src/docfx/lib/log/Telemetry.cs:96]: nit: why not directly call `s_telemetryClient.TrackEvent`? this helper method does not seem to add much value.

Comment (928PJY) [reply src/docfx/lib/log/Telemetry.cs:94]: this method is used to construct the eventTelemetry?

---
In reply to: [452600406](https://github.com/dotnet/docfx/pull/6216#discussion_r452600406) [](ancestors = 452600406)

File: src/docfx/lib/log/Telemetry.cs [hunk 3]
```diff
@@ -147,5 +184,50 @@ private static string CoalesceEmpty(string? str)
...
+        private static string SerializeConfigObjectWithMaxLength(JObject graph)
+        {
+            TryRemoveNestedObject(graph);
+            var result = JsonUtility.Serialize(graph);
+            if (result.Length <= MaxEventPropertyLength)
+            {
+                return result;
+            }
+
+            TryRemoveLongArray(graph);
```
Comment (live1206) [inline src/docfx/lib/log/Telemetry.cs:197]: Feel a bit strange to trim again here and also why not trim JObject as well, how about reduce `MaxChildrenLength` instead?

File: src/docfx/lib/log/Telemetry.cs [hunk 4]
```diff
@@ -73,6 +78,22 @@ public static void SetRepository(string? repo, string? branch)
...
+            var hashCode = HashUtility.GetMd5Hash(docfxConfigTelemetryValue);
+            if (docfxConfigTelemetryValue.Length > MaxEventPropertyLength)
+            {
+                docfxConfigTelemetryValue = SerializeConfigObjectWithMaxLength(docfxConfig);
+            }
+
+            var dimensions = new Dictionary<string, string>();
+            dimensions["DocsetName"] = docsetName;
+            dimensions["Config"] = docfxConfigTelemetryValue;
+            dimensions["ContentHash"] = hashCode;
```
Comment (yufeih) [inline src/docfx/lib/log/Telemetry.cs:93]: What is the purpose of this hash?

Comment (928PJY) [reply src/docfx/lib/log/Telemetry.cs:93]: `hashcode` will be used by the Feature Utilization team to detect the config change once we remove some nested object from the too-long config

File: src/docfx/lib/log/Telemetry.cs (hunk 3 above)
Comment (928PJY) [reply src/docfx/lib/log/Telemetry.cs]: remove it then

File: src/docfx/lib/log/Telemetry.cs [hunk 5]
```diff
@@ -73,6 +78,24 @@ public static void SetRepository(string? repo, string? branch)
...
         }
 
+        public static void TrackDocfxConfig(string fileName, string docsetName, JObject docfxConfig)
+        {
+            var newValue = JsonUtility.DeepClone(docfxConfig) as JObject;
+            var docfxConfigTelemetryValue = JsonUtility.Serialize(newValue!);
+            var hashCode = HashUtility.GetMd5Hash(docfxConfigTelemetryValue);
+            if (docfxConfigTelemetryValue.Length > MaxEventPropertyLength)
+            {
+                TryRemoveNestedObject(newValue!);
```
Comment (yufeih) [inline src/docfx/lib/log/Telemetry.cs:88]: The `!` here will cause an `NRE` if the config is not a `JObject`, in which case we can skip the fallback.

File: src/docfx/lib/log/Telemetry.cs [hunk 6]
```diff
@@ -73,6 +78,24 @@ public static void SetRepository(string? repo, string? branch)
             s_telemetryClient.Context.GlobalProperties["Branch"] = s_branch;
         }
 
+        public static void TrackDocfxConfig(string docsetName, JObject docfxConfig)
+        {
+            var docfxConfigTelemetryValue = JsonUtility.Serialize(docfxConfig);
+            var hashCode = HashUtility.GetMd5Hash(docfxConfigTelemetryValue);
+            if (docfxConfigTelemetryValue.Length > MaxEventPropertyLength)
+            {
+                var newValue = JsonUtility.DeepClone(docfxConfig) as JObject;
```
Comment (yufeih) [inline src/docfx/lib/log/Telemetry.cs:87]:
```suggestion
                var newValue = (JObject)JsonUtility.DeepClone(docfxConfig);
```

**Comment by KalleOlaviNiemitalo** [comment]
Does this now send the whole docfx.json to Application Insights? How will I be able to opt out of that when building proprietary non-Microsoft documentation?

**Comment by yufeih** [comment]
@KalleOlaviNiemitalo Today you have to opt in to collect telemetry using the CLI (by manually setting the AI instrumentation key) and we only collect information Microsoft docs repos.

@928PJY @herohua , if in the future we choose to collect telemetry by default to the public (whether in CLI or vscode extension), we'll need a privacy review.

**Pull request merged**